### ISSUE TITLE
feature request: handle multivariate (multi-dimension) time-series

### ISSUE THREAD

**Issue by mw66** [garrettwrong/cuTWED#26]
e.g. as in:

https://marcocuturi.net/GA.html

TGA_python_wrapper.

BTW, anyone know how TWED performs compared with Triangular Global Alignment Kernels on the above page?

Since both metrics said to satisfy triangular inequality, which metric is better?

**Comment by garrettwrong** [comment]
I thought this code already ran for arbitrary dimensions, though I haven't used it much in that regard. The original reference code by Marteau was limited, but only required a few lines of code to change... If this isn't exposed it was just an oversight. It isn't a limitation of the method as I understand it... We just compute distances in R^d... remainder of algorithm is the same.

I'll take a closer look before I comment further in case I misunderstood.  I'm not familiar with that software/method. I'll check it out.

**Comment by mw66** [comment]
> I thought this code already ran for arbitrary dimensions

Can you add some demo / test code for multi-dimension?

**Comment by garrettwrong** [comment]
There is a unit test which uses some MNIST 4x4 data as an `R^4` by 4 time sample data set. The test checks for consistency between the raw Python implementation on the Wiki, my `R^N` extension of Marteau's C code (`ctwed`), and `cuTWED`.

https://github.com/garrettwrong/cuTWED/blob/master/tests/test_Rn.py

The unit test may be a little too terse for an example so I will try to add a more straightforward example in a notebook (with better data).

**Comment by garrettwrong** [comment]
@mingwugmail , I've created a notebook that downloads and extracts a real multivariate data set used in other machine learning reference works. The data is 3D (x,y,pressure) with varying time series lengths representing measured hand written character trajectories.  I am posting a Google Colab notebook which should run live in a browser.  (You may need to set the backend to GPU accelerator mode, I'm not sure if the settings will carry over on sharing...).

https://colab.research.google.com/drive/128LVV4-Ho4vnG_b8oUiyR9JivHyL098t?usp=sharing

In case you have any trouble with Google Colab, I attached a PDF.
[Multivariate_cuTWED.ipynb - Colaboratory.pdf](https://github.com/garrettwrong/cuTWED/files/6152438/Multivariate_cuTWED.ipynb.-.Colaboratory.pdf)

I hope that helps, and if it satisfies your request for R^N please close the issue for me.

I will look at the TGA stuff when I get the chance. Thanks!

**Comment by mw66** [comment]
Thank you.

BTW, on the TGA page, there is an embeded youtube video comparing DTW vs Global Alignment Kernel:

https://www.youtube.com/watch?v=311olQkfFp0

(somehow it only showing on smart phones, not on desktop browser).

**Comment by garrettwrong** [comment]
Neat. I can say that the performance of TWED has some different character than DTW, but I like the clip.  Some time ago I was referred to the paper below that compares many approaches.  In those tests TWED outperformed DTW (and mostly everything else...) Maybe it is useful to you too.

"Debunking Four Long-Standing Misconceptions of Time-Series Distance Measures" Paparrizos et al

**Comment by mw66** [comment]
Thanks for the paper, actually I found they did include GAK in their study:

the Global Alignment Kernel (GAK) [38] 

Dynamic Time Warping Kernel (KDTW) [93], and I found it here: https://github.com/pfmarteau/KDTW

TimeWarp Edit (TWE) distance [92] is a measure that combinesmerits from LCSS and DTW.

(which you implemented here).

And in their paper on Figure 9: Accuracy-to-runtime comparison, TWE indeed have the highest accuracy.

[38]Marco Cuturi. 2011.  Fast global alignment kernels. InProceedingsof the 28th international conference on machine learning (ICML-11).929–936

[92]Pierre-François Marteau. 2008. Time warp edit distance with stiffnessadjustment for time series matching.IEEE Transactions on PatternAnalysis and Machine Intelligence31, 2 (2008), 306–318.

[93]Pierre-François Marteau and Sylvie Gibet. 2014.  On recursive editdistance kernels with application to time series classification.IEEEtransactions on neural networks and learning systems26, 6 (2014),1121–1133
